### DISCUSSION TITLE
Cloudfying Self-hosted VMs

### DISCUSSION THREAD

**Discussion by turkmenkaan** [ubicloud/ubicloud#1754] [Ideas]
Not sure if this is something being planned, or already exists but I wasn't able to find anything about it yet. Does Ubicloud support cloudfying on-prem machines i.e. running a VM (in proxmox in this case) and using the control plane to deploy services on the running VM?

**Comment by byucesoy** [comment]
Hi @turkmenkaan,

Currently we only support cloudifying bare metal servers. Nested virtualization is not something we tried and we don't have any plans to work on it any time soon.

**Comment by aparcar** [comment]
@byucesoy could you give me some pointers of what's missing to support that? I'm mostly interested in very small VMs doing tiny jobs, which should still be running in separate QEMU VM.

**Reply by fdr** [reply]
mostly, you need 

1) nested virtualization to work there are many writings about this on the internet, there's nothing particularly special about Ubicloud's stack for this. Substantially, that means that when cloud-hypervisor is started it should...not crash.

2) A networking situation that is compatible.

The second is a bit more ubicloud-specific. Generally, we rely on chunks of IPv6 address space delegated to us, and slice it up even more for each VM on a host. We largely platform on Hetzner, so the most mature situation is, each computer this control plane SSHs into gets its own /64 that is statically routed, and then we randomly find /79 blocks within for each VM as it comes and go. Of that /79, it broken down into two /80s, one being delegated to the VM and another for saved for internal system communication, such as for private networking. You might find yourself wanting to do /70 and /95, or whatever. The system doeesn't work like this, but by messing with some constants, it should work (in fact, I changed the size a long time ago, 11c2229c51b5198e97ea5458250653cd4dc5eeb9)

Generally, we prefer the traffic be "statically routed," to avoid having to use NDP (IPv6's ARP-like thing) to tell the switch about every single address that might be listening to traffic. There's a little `ndp_needed` flag tucked away (and barely tested) when we deal with such networks, generally in a prototyping capacity.

In the context of nested virtualization, you'd want to forward most traffic to the VM that is intended to be managed by Ubicloud, e.g. if you have a /64 then take a /65 inside that, and make a route to the VM, and adjust `learn_network.rb` to pick up the configuration properly, or etch correct data manually into the VmHost record when it's created.

You might also need to mess with the subnet slicing arithmetic in `random_private_ipv6` to deal with smaller ranges. We have not expended complications supporting prefix size situations outside our direct experience.

If you want IPv4 support, you'll also have to forward that traffic to the VM hosting the VMs set up by ubicloud. This one is in some ways simpler, because IPv4 is not used for private networking. You can see `hetznerify` which uses the API to fill in some records, you'd have to create those same records for your situation (or write a proxmox client, or whatever, sounds interesting)
